File: MineInAbyss-API/src/main/java/com/derongan/minecraft/mineinabyss/World/Layer.java
package com.derongan.minecraft.mineinabyss.World;

import com.derongan.minecraft.mineinabyss.Ascension.Effect.AscensionEffect;
import com.derongan.minecraft.mineinabyss.Ascension.Effect.AscensionEffectBuilder;
import org.bukkit.World;

import java.util.List;

public interface Layer {
    /**
     * Gets the name of this Layer. This name may not match the worldname.
     * @return The name of the layer
     */
    String getName();

    /**
     * Get the sub header
     */
    String getSub();

    /**
     * Gets the index of this Layer. Higher layers have lower index
     * @return The index of this layer
     */
    int getIndex();

    /**
     * Gets the sections in this layer. This list is immutable.
     * @return An immutable list containing the sections of this layer
     */
    List<Section> getSections();

    /**
     * Gets the section at index or throws //todo what to throw
     * @param index The index of the section
     * @return The section at the index
     */
    Section getSection(int index);

    /**
     * Gets the effects of ascending on this layer
     */
    List<AscensionEffectBuilder> getAscensionEffects();


    /**
     * Get custom death message suffix for this layer
     * @return The custom death suffix to use
     */
    default String getDeathMessage(){
        return " in the depths of the abyss";
    }
}

File: src/main/java/com/derongan/minecraft/mineinabyss/world/Layer.kt
package com.derongan.minecraft.mineinabyss.world

import com.derongan.minecraft.deeperworld.world.section.Section
import com.derongan.minecraft.mineinabyss.ascension.effect.AscensionEffectBuilder

/**
 * @property name the name of this Layer. This name may not match the world name.
 * @property sub The sub header.
 * @property sections The sections in this layer. This list is immutable.
 * @property ascensionEffects The effects of ascending on this layer.
 * @property maxCurseRadius The radius at which the curse has maximum effect.
 * @property minCurseRadius The radius at which the curse has minimum effect.
 * @property minCurseMultiplier Effectively multiply the distance you travel up by this amount when you are closer than
 * minCurseRadius away.
 * @property maxCurseMultiplier Effectively multiply the distance you travel up by this amount when you are further than
 * maxCurseRadius away.
 * @property curseOverrideRegions Regions that have custom curse levels.
 * @property startDepth Starting depth of this layer.
 * @property endDepth End depth of this layer.
 * @property deathMessage Custom death message suffix for this Layer.
 */
interface Layer {
    val name: String?
    val sub: String?
    val deathMessage: String? get() = " in the depths of the abyss"
    val ascensionEffects: List<AscensionEffectBuilder<*>>
    val maxCurseRadius: Float
    val minCurseRadius: Float
    val minCurseMultiplier: Float
    val maxCurseMultiplier: Float
    val curseOverrideRegions: List<CurseRegion>
    val startDepth: Int
    val endDepth: Int
    val sections: List<Section>

    /** Returns whether or not this layer contains this section */
    fun containsSection(section: Section): Boolean

}